src/core/qgsdatadefined.cpp

```diff
@@ -43,6 +43,18 @@ QgsDataDefined::QgsDataDefined( const QgsExpression * expression )
   mExpressionPrepared = false;
 }
 
+QgsDataDefined::QgsDataDefined( const QString & expressionString )
+{
+  QgsExpression expression( expressionString );
+  mActive = expression.rootNode();
+  mUseExpression = expression.rootNode() && !dynamic_cast<const QgsExpression::NodeColumnRef*>( expression.rootNode() );
```

Comment:
I'd suggest moving this check to something like QgsExpression::isSingleField() , and adding a few unit tests for it.